frontend/app-development/features/dataModelling/SchemaEditorWithToolbar/TopToolbar/CreateNewWrapper.tsx

```diff
@@ -1,110 +1,88 @@
+import type { ChangeEvent, KeyboardEvent } from 'react';
 import React, { useState } from 'react';
 import classes from './CreateNewWrapper.module.css';
-import { ErrorMessage, Textfield } from '@digdir/designsystemet-react';
 import { useTranslation } from 'react-i18next';
 import { PlusIcon } from '@studio/icons';
-import { extractModelNamesFromMetadataList } from '../../../../utils/metadataUtils';
 import type { DataModelMetadata } from 'app-shared/types/DataModelMetadata';
-import { StudioButton, StudioPopover } from '@studio/components';
-import { useStudioEnvironmentParams } from 'app-shared/hooks/useStudioEnvironmentParams';
-import { useAppMetadataQuery } from 'app-shared/hooks/queries';
+import { StudioButton, StudioPopover, StudioTextfield } from '@studio/components';
+import { useValidateSchemaName } from '../../hooks/useValidateSchemaName';
+import { useCreateDataModelMutation } from '../../../../hooks/mutations';
 
 export interface CreateNewWrapperProps {
   disabled: boolean;
-  createNewOpen: boolean;
+  isCreateNewOpen: boolean;
   createPathOption?: boolean;
   dataModels: DataModelMetadata[];
-  setCreateNewOpen: (open: boolean) => void;
-  handleCreateSchema: (props: { name: string; relativePath: string | undefined }) => void;
+  setIsCreateNewOpen: (open: boolean) => void;
 }
 
 export function CreateNewWrapper({
   disabled,
   createPathOption = false,
-  createNewOpen,
+  isCreateNewOpen,
   dataModels,
-  setCreateNewOpen,
-  handleCreateSchema,
+  setIsCreateNewOpen,
 }: CreateNewWrapperProps) {
   const { t } = useTranslation();
   const [newModelName, setNewModelName] = useState('');
-  const [nameError, setNameError] = useState('');
-
-  const { org, app } = useStudioEnvironmentParams();
-  const { data: appMetadata } = useAppMetadataQuery(org, app);
-  const modelNames = extractModelNamesFromMetadataList(dataModels);
+  const { validateName, nameError, setNameError } = useValidateSchemaName(dataModels);
```

Comment:
Good investigation! 🕵 Maybe the issue you mention in alt 2 can be solved with setting the cache in useAppMetadataMutation instead of invalidating? 🤔

Reply:
What do you mean by setting the cache? I'm not very familiar with this 😅

Reply:
I meant something like this:
```
return useMutation({
    mutationFn: (payload: ApplicationMetadata) => updateAppMetadata(org, app, payload),
    onSuccess: (payload) => {
      queryClient.setQueryData( [QueryKey.AppMetadata, org, app], payload);
    },
  });
```  
But I see that that would not work since we are not directly modifying the appMetadata in frontend.
On the other hand, I see that we actually do not invalidate the appMetadata at all when creating a data model. You can try to add this line:
```
queryClient.invalidateQueries({ queryKey: [QueryKey.AppMetadata, org, app] }),
```
in line 28 in `useCreateDataModelMutation` 🙈

Reply:
Hmm, it seems like `appMetaData` is only updated when Studio sees that an XSD file has been made, and that only happens after pressing "Generate datamodels", even if we invalidate the `appMetaData` query in `useCreateDataModelMutation`.

We could just call `useDataModelsJsonQuery` in `useValidationSchema` to avoid this issue, but then I think we should also call `useDataModelsXsdQuery` just to be safe. Like this: 
```tsx
export const useValidateSchemaName = () => {
  const [nameError, setNameError] = useState('');
  const { org, app } = useStudioEnvironmentParams();
  const { data: appMetadata } = useAppMetadataQuery(org, app);
  const { data: jsonDataModels } = useDataModelsJsonQuery(org, app);
  const { data: xsdDataModels } = useDataModelsXsdQuery(org, app);
  const { t } = useTranslation();

  const dataModels = mergeJsonAndXsdData(jsonDataModels, xsdDataModels);

  const validateName = (name: string): void => {
...
```
But then we have code duplication with the `DataModelling` component, and can maybe opt for the hook in alternative 2 above?

Reply:
No worries! I've got a better grasp on React Query because of this. And Lars is going to use `useValidateSchemaName` in Utforming, so the hook will be easier to reuse when its not dependent on the `dataModels` parameter 😊 

I did an experiment with and without memoization, and you are correct: `mergeJsonAndXsdData` will be executed for every single letter we type on the keyboard. Wrapping the function in `useMemo` fixed this.

Reply:
As we discussed on Slack, I have now removed fetching of data models and application metadata from `useValidateSchemaName`, so that it is not dependent on anything from `frontend/app-development/features/dataModelling`. `useValidateSchemaName` now lives in `frontend/packages/shared` 😊